scipy/special/xsf/numpy.h

```diff
@@ -14,14 +14,37 @@
 
 #include <numpy/arrayobject.h>
 #include <numpy/npy_3kcompat.h>
+#include <numpy/npy_math.h>
 #include <numpy/ufuncobject.h>
 
 #include "dual.h"
-#include "sf_error.h"
+#include "error.h"
```

Comment:
All names in `error.h` use `SF_`, as does the comment below. I'd probably keep the `sf_error.h` name (also easier to search for, `error.h` is very generic).

Reply:
Yeah, I think that makes sense. Calling the error handling header `error.h` in what is now `xsf` actually goes back to the very beginning of this project, see https://github.com/scipy/scipy/pull/19482. If I remember correctly, @izaid chose the name, but I can't recall his rationale for it. Maybe he'll be able to explain.

Reply:
`xsf_error.h`?

(back to the peanut gallery)

Reply:
@rgommers, I spoke to @izaid, and we think you've pointed out something valid, but that the conclusion should be reversed. Instead of renaming `error.h` to `sf_error.h`, what we should do instead is stop using the `SF_` prefix within `xsf`. Instead of defining the `enum`

```C++
typedef enum {
    SF_ERROR_OK = 0,    /* no error /
    SF_ERROR_SINGULAR,  / singularity encountered /
    SF_ERROR_UNDERFLOW, / floating point underflow /
    SF_ERROR_OVERFLOW,  / floating point overflow /
    SF_ERROR_SLOW,      / too many iterations required /
    SF_ERROR_LOSS,      / loss of precision /
    SF_ERROR_NO_RESULT, / no result obtained /
    SF_ERROR_DOMAIN,    / out of domain /
    SF_ERROR_ARG,       / invalid input parameter /
    SF_ERROR_OTHER,     / unclassified error */
    SF_ERROR__LAST
} sf_error_t;
```

using the prefixes, we should use  an `enum class` in a namespace like this

```C++
namespace xsf {

enum class Error {
    OK = 0,           // no error
    Singular,         // singularity encountered
    Underflow,        // floating point underflow
    Overflow,         // floating point overflow
    Slow,             // too many iterations required
    Loss,             // loss of precision
    NoResult,         // no result obtained
    Domain,           // out of domain
    Arg,              // invalid input parameter
    Other,            // unclassified error
    _Last
};
```

and then something like `SF_ERROR_SINGULAR` would be replaced with `xsf::Error::Singular`. 

The diff when making this change will be large though, since `set_error` appears many times, and the change would need to be made for all such appearances. My preference would be to punt on this until after the `xsf` migration.

Reply:
> `xsf_error.h`?
> 
> (back to the peanut gallery)

I missed this comment, but Irwin explained to me that having a prefix like this in a header name would be very unusual in a C++ library. I guess that the header is in `xsf/error.h` and it's contents live in the `xsf` namespace should be enough to make things clear.

It looks like this for instance when it gets included outside of the `xsf` header library

https://github.com/scipy/scipy/blob/35ad5912fa6901f1d090596224db31f824284b7d/scipy/special/sf_error.cc#L9

Reply:
> and then something like `SF_ERROR_SINGULAR` would be replaced with `xsf::Error::Singular`.
> 
> The diff when making this change will be large though, since `set_error` appears many times, and the change would need to be made for all such appearances. My preference would be to punt on this until after the `xsf` migration.

As long as naming is consistent and works, any naming choice is fine with me.

The shared library is also used from C code, so changing the `typedef enum` to `enum class Error` may see a hiccup there perhaps? (if so, you'll bump into it soon enough anyway)

Reply:
> having a prefix like this in a header name would be very unusual in a C++ library.

Not really. It is actually a good practice so that linkers don't get confused when multiple sources are linked to a single binary. This happens very often and if you want this code to end up in complicated places it is very typical to prefix everything.

Reply:
> > having a prefix like this in a header name would be very unusual in a C++ library.
> 
> Not really. It is actually a good practice so that linkers don't get confused when multiple sources are linked to a single binary. This happens very often and if you want this code to end up in complicated places it is very typical to prefix everything. Filenames don't make any difference within the same codebase but do matter when it is mixed up with other code.

Irwin's busy now, and I'm really just passing on what he told me. I did a brief survey of libraries I know about and found:

Boost doesn't prefix (I already knew this), [include/boost/algorithm/string.hpp](https://github.com/boostorg/algorithm/blob/32c5a6327cfdca5d41ce0f1d8849b811886daa2f/include/boost/algorithm/string.hpp)
Kokkos does prefix, [algorithms/src/Kokkos_Sort.hpp](https://github.com/kokkos/kokkos/blob/develop/algorithms/src/Kokkos_Sort.hpp)
Eigen does not prefix, [Eigen/src/Core/Array.h](https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/Array.h?ref_type=heads)
Blaze does not prefix, [blaze/Math.h](https://bitbucket.org/blaze-lib/blaze/src/master/blaze/Math.h)
TNT does prefix, [TNT/tnt_array1d.h](https://github.com/dvalters/Template-Numerical-Toolkit/blob/master/TNT/tnt_array1d.h)

A mix of prefixers and non-prefixers. The use of namespaces should prevent symbol conflicts, and I think as long as every external project which includes something from `xsf` includes headers as `xsf/<some-header>.h`, then the linker shouldn't get confused. I will walk back prefixing being unusual since 2/5 of the projects I checked do prefix.